include/oxen/quic/ip.hpp

```diff
@@ -160,48 +147,124 @@ namespace oxen::quic
+            ipv4 ip;
+            uint8_t mask;
+
+          protected:
+            constexpr masked_ipv4() = default;
+            constexpr masked_ipv4(ipv4 b, uint8_t m) : ip{b}, mask{m} {}
+
+          public:
+            constexpr ipv4 max_ip() const
+            {
+                auto b = ip.to_base(mask);
 
-        constexpr ipv6 max_ip() const
+                if (mask < 32)
+                    b.addr |= (uint32_t{1} << (32 - mask)) - 1;
+
+                return b;
+            }
+
+            constexpr auto operator<=>(const masked_ipv4& other) const
+            {
+                return std::tie(ip, mask) <=> std::tie(other.ip, other.mask);
+            }
+
+            constexpr bool operator==(const masked_ipv4& other) const { return (*this <=> other) == 0; }
+            constexpr bool contains(const ipv4& other) const { return other.to_base(mask) == ip; }
+
+            std::string to_string() const;
+            constexpr static bool to_string_formattable = true;
+        };
+
+        struct masked_ipv6
         {
-            auto b = base.to_base(mask);
+            ipv6 ip;
+            uint8_t mask;
 
-            if (mask > 64)
+          protected:
+            constexpr masked_ipv6() = default;
+            constexpr masked_ipv6(ipv6 b, uint8_t m) : ip{b}, mask{m} {}
+
+          public:
+            constexpr ipv6 max_ip() const
             {
-                b.hi = base.hi;
-                b.lo |= (uint64_t{1} << (128 - mask)) - 1;
+                auto b = ip.to_base(mask);
+
+                if (mask > 64)
+                {
+                    b.hi = ip.hi;
+                    b.lo |= (uint64_t{1} << (128 - mask)) - 1;
+                }
+                else
+                {
+                    b.hi |= (uint64_t{1} << (64 - mask)) - 1;
+                    b.lo = ~uint64_t{0};
+                }
+
+                return b;
             }
-            else
+
+            constexpr auto operator<=>(const masked_ipv6& other) const
             {
-                b.hi |= (uint64_t{1} << (64 - mask)) - 1;
-                b.lo = ~uint64_t{0};
+                return std::tie(ip, mask) <=> std::tie(other.ip, other.mask);
             }
 
-            return b;
-        }
+            constexpr bool operator==(const masked_ipv6& other) const { return (*this <=> other) == 0; }
+            constexpr bool contains(const ipv6& other) const { return other.to_base(mask) == ip; }
 
-        constexpr ipv6_net() = default;
+            std::string to_string() const;
+            constexpr static bool to_string_formattable = true;
+        };
+    }  //  namespace detail
+
+    struct ipv4_net : public detail::masked_ipv4
+    {
+        constexpr ipv4_net() = default;
+        constexpr ipv4_net(ipv4 b, uint8_t m) : detail::masked_ipv4{b.to_base(m), m} {}
+    };
 
-        constexpr ipv6_net(ipv6 b, uint8_t m) : base{b.to_base(m)}, mask{m} {}
+    inline constexpr ipv4_net operator/(const ipv4& a, uint8_t mask)
```

Comment:
I think that this operator should return a ipv4_range now instead of a _net, as I think it's a bit more natural for ipv4{1,2,3,4} / 24 to mean an IP with a netmask, rather than the 1.2.3.0/24 network range itself.

Reply:
I do realize that gives us no nice operator for representing a base network, though.

That's not the end of the world, but perhaps we could use another operator, such as `&`, for it?

ipv4{127.0.0.1} / 8 == ip with mask, CIDR notation 127.0.0.1/8
ipv4{127.0.0.1} & 8 == ipv4 subnet, CIDR notation 127.0.0.0/8